File: src/utils/offline-queue.ts
/**
 * Cola de sincronización offline.
 * Almacena operaciones pendientes cuando no hay conexión.
 * Máximo 50 elementos. Sincronización FIFO.
 */
import AsyncStorage from '@react-native-async-storage/async-storage';
import { SyncItem, SyncItemType } from '../types';

const QUEUE_KEY = 'evora_offline_queue';
const MAX_ITEMS = 50;
const MAX_RETRIES = 3;

export const offlineQueue = {
  /**
   * Añadir elemento a la cola.
   */
  async enqueue(type: SyncItemType, data: unknown): Promise<boolean> {
    const queue = await this.getAll();

    if (queue.length >= MAX_ITEMS) {
      return false; // Cola llena
    }

    const item: SyncItem = {
      id: `${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
      type,
      status: 'pending',
      data,
      created_at: new Date().toISOString(),
      retries: 0,
    };

    queue.push(item);
    await AsyncStorage.setItem(QUEUE_KEY, JSON.stringify(queue));
    return true;
  },

  /**
   * Obtener todos los elementos de la cola.
   */
  async getAll(): Promise<SyncItem[]> {
    const raw = await AsyncStorage.getItem(QUEUE_KEY);
    return raw ? JSON.parse(raw) : [];
  },

  /**
   * Obtener número de elementos pendientes.
   */
  async getPendingCount(): Promise<number> {
    const queue = await this.getAll();
    return queue.filter((item) => item.status === 'pending').length;
  },

  /**
   * Marcar elemento como en proceso de sincronización.
   */
  async markSyncing(id: string): Promise<void> {
    await this.updateStatus(id, 'syncing');
  },

  /**
   * Eliminar elemento sincronizado correctamente.
   */
  async remove(id: string): Promise<void> {
    const queue = await this.getAll();
    const filtered = queue.filter((item) => item.id !== id);
    await AsyncStorage.setItem(QUEUE_KEY, JSON.stringify(filtered));
  },

  /**
   * Marcar elemento como fallido. Si supera MAX_RETRIES, queda en error.
   */
  async markFailed(id: string): Promise<void> {
    const queue = await this.getAll();
    const item = queue.find((i) => i.id === id);
    if (item) {
      item.retries += 1;
      item.status = item.retries >= MAX_RETRIES ? 'error' : 'pending';
    }
    await AsyncStorage.setItem(QUEUE_KEY, JSON.stringify(queue));
  },

  /**
   * Limpiar toda la cola.
   */
  async clear(): Promise<void> {
    await AsyncStorage.removeItem(QUEUE_KEY);
  },

  // --- Helpers privados ---

  async updateStatus(id: string, status: SyncItem['status']): Promise<void> {
    const queue = await this.getAll();
    const item = queue.find((i) => i.id === id);
    if (item) {
      item.status = status;
    }
    await AsyncStorage.setItem(QUEUE_KEY, JSON.stringify(queue));
  },
};

File: lib/core/offline_queue.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';
import '../models/sync_item.dart';

/// Cola de sincronización offline.
/// Almacena operaciones pendientes cuando no hay conexión.
/// Máximo 50 elementos. Sincronización FIFO.
class OfflineQueue {
  static const _queueKey = 'evora_offline_queue';
  static const _maxItems = 50;
  static const maxRetries = 3;
  static const _uuid = Uuid();

  /// Añadir elemento a la cola.
  static Future<bool> enqueue(SyncItemType type, Map<String, dynamic> data) async {
    final queue = await getAll();
    if (queue.length >= _maxItems) return false;

    final item = SyncItem(
      id: '${DateTime.now().millisecondsSinceEpoch}-${_uuid.v4().substring(0, 7)}',
      type: type,
      status: SyncItemStatus.pending,
      data: data,
      createdAt: DateTime.now().toIso8601String(),
    );

    queue.add(item);
    await _save(queue);
    return true;
  }

  /// Obtener todos los elementos de la cola.
  static Future<List<SyncItem>> getAll() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_queueKey);
    if (raw == null) return [];
    final list = jsonDecode(raw) as List<dynamic>;
    return list
        .map((e) => SyncItem.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Obtener número de elementos pendientes.
  static Future<int> getPendingCount() async {
    final queue = await getAll();
    return queue.where((item) => item.status == SyncItemStatus.pending).length;
  }

  /// Marcar elemento como en proceso de sincronización.
  static Future<void> markSyncing(String id) async {
    await _updateStatus(id, SyncItemStatus.syncing);
  }

  /// Eliminar elemento sincronizado correctamente.
  static Future<void> remove(String id) async {
    final queue = await getAll();
    queue.removeWhere((item) => item.id == id);
    await _save(queue);
  }

  /// Marcar elemento como fallido. Si supera maxRetries, queda en error.
  static Future<void> markFailed(String id) async {
    final queue = await getAll();
    final item = queue.firstWhere((i) => i.id == id, orElse: () => queue.first);
    if (item.id == id) {
      item.retries += 1;
      item.status =
          item.retries >= maxRetries ? SyncItemStatus.error : SyncItemStatus.pending;
    }
    await _save(queue);
  }

  /// Resetear elementos con error a pendiente.
  static Future<void> resetErrors() async {
    final queue = await getAll();
    for (final item in queue) {
      if (item.status == SyncItemStatus.error) {
        item.status = SyncItemStatus.pending;
        item.retries = 0;
      }
    }
    await _save(queue);
  }

  /// Limpiar toda la cola.
  static Future<void> clear() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_queueKey);
  }

  static Future<void> _updateStatus(String id, SyncItemStatus status) async {
    final queue = await getAll();
    for (final item in queue) {
      if (item.id == id) {
        item.status = status;
        break;
      }
    }
    await _save(queue);
  }

  static Future<void> _save(List<SyncItem> queue) async {
    final prefs = await SharedPreferences.getInstance();
    final json = jsonEncode(queue.map((e) => e.toJson()).toList());
    await prefs.setString(_queueKey, json);
  }
}
